File: packages/ballerina-central/src/result.ts
/*
 * Copyright (c) 2026, WSO2 LLC. (https://www.wso2.com).
 *
 * WSO2 LLC. licenses this file to you under the Apache License,
 * Version 2.0 (the "License"); you may not use this file except
 * in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied. See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */

/**
 * Failures are values, not thrown classes.
 *
 * The reader has exactly one caller shape — a command whose stdout is a
 * Ballerina file and whose stderr is one JSON object — so the question every
 * failure has to answer is "what does the agent reading this do next". A
 * discriminated union answers it exhaustively: `exitCodeFor` switches over
 * `kind` with a `never` fallthrough, so a new failure mode fails the build
 * until someone decides what it costs the caller.
 */

/** One place a payload stopped matching the schema. */
export interface SchemaIssue {
  /** Dotted path into the Central payload, e.g. `docsData.modules.0.records`. */
  readonly path: string;
  readonly message: string;
}

export type Failure =
  /** The caller's arguments are wrong — nothing upstream was contacted. */
  | { readonly kind: "validation"; readonly message: string; readonly suggestion: string }
  /** Central has no such package, or no such version of it. */
  | { readonly kind: "package-not-found"; readonly qualified: string; readonly suggestion: string }
  /** Central answered, but not usefully — a 4xx/5xx, a network error, a bad body. */
  | {
      readonly kind: "upstream";
      readonly url: string;
      readonly attempts: number;
      readonly message: string;
      readonly status?: number;
    }
  /** Central did not answer inside the budget. */
  | { readonly kind: "timeout"; readonly url: string; readonly budgetMs: number }
  /** Central answered with a shape this reader does not understand. */
  | { readonly kind: "schema-drift"; readonly qualified: string; readonly issues: readonly SchemaIssue[] }
  /** The package is fine, but no module of it carries a written guide. */
  | { readonly kind: "no-readme"; readonly qualified: string; readonly suggestion: string };

export type Result<T> = { readonly ok: true; readonly value: T } | { readonly ok: false; readonly error: Failure };

export function ok<T>(value: T): Result<T> {
  return { ok: true, value };
}

export function err<T = never>(error: Failure): Result<T> {
  return { ok: false, error };
}

/**
 * Process exit code for a failure.
 *
 * Two codes, because the caller has two responses: a `validation` failure is
 * the caller's own mistake and re-running unchanged cannot help (exit 2, same
 * as a missing argument), while everything else is a fact about Central, the
 * network or the package itself, where running again could legitimately give a
 * different answer (exit 1). Each failure carries its own `suggestion` for what
 * to do about it; none of them is a licence to guess at a signature instead.
 */
export function exitCodeFor(failure: Failure): 1 | 2 {
  switch (failure.kind) {
    case "validation":
      return 2;
    case "package-not-found":
    case "upstream":
    case "timeout":
    case "schema-drift":
    case "no-readme":
      return 1;
    default: {
      const exhaustive: never = failure;
      return exhaustive;
    }
  }
}

/**
 * The one line a failing run writes to stderr. Kept as a single JSON object so
 * an agent can read it without a parser and a human can read it without tools;
 * `kind` is the field worth branching on.
 */
export function describeFailure(failure: Failure): string {
  return JSON.stringify(failure);
}

File: packages/bal-library-tool/native/src/main/java/io/ballerina/library/Result.java
/*
 *  Copyright (c) 2026, WSO2 LLC. (http://www.wso2.com)
 *
 *  WSO2 LLC. licenses this file to you under the Apache License,
 *  Version 2.0 (the "License"); you may not use this file except
 *  in compliance with the License.
 *  You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 *  Unless required by applicable law or agreed to in writing,
 *  software distributed under the License is distributed on an
 *  "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 *  KIND, either express or implied.  See the License for the
 *  specific language governing permissions and limitations
 *  under the License.
 */

package io.ballerina.library;

/**
 * A value or a {@link Failure}, never an exception.
 *
 * <p>The reader has exactly one caller shape — a command whose stdout is a document and whose stderr
 * is one JSON object — so a network hiccup, a bad argument and a payload we cannot parse are all
 * values that travel back to the one place that knows what to do with them. Nothing in the render
 * pipeline throws.
 *
 * @param <T> the value a successful result carries
 * @since 0.1.0
 */
public sealed interface Result<T> {

    /** A value. */
    record Ok<T>(T value) implements Result<T> { }

    /** A failure. The type parameter is phantom, so an error can be re-typed by {@link #cast()}. */
    record Err<T>(Failure failure) implements Result<T> { }

    static <T> Result<T> ok(T value) {
        return new Ok<>(value);
    }

    static <T> Result<T> err(Failure failure) {
        return new Err<>(failure);
    }

    default boolean isOk() {
        return this instanceof Ok<T>;
    }

    /**
     * The value. Only legal on an {@code Ok}, which is the discipline every call site follows by
     * checking {@link #isOk()} first.
     */
    default T value() {
        return switch (this) {
            case Ok<T> ok -> ok.value();
            case Err<T> err -> throw new IllegalStateException("not ok: " + err.failure());
        };
    }

    /** The failure. Only legal on an {@code Err}. */
    default Failure failure() {
        return switch (this) {
            case Ok<T> ok -> throw new IllegalStateException("not a failure: " + ok.value());
            case Err<T> err -> err.failure();
        };
    }

    /** The same failure at another value type, so a guard clause can return it unchanged. */
    default <U> Result<U> cast() {
        return switch (this) {
            case Ok<T> ok -> throw new IllegalStateException("only a failure can be re-typed");
            case Err<T> err -> Result.err(err.failure());
        };
    }

}
